os/mm/mm_heap/mm_mallinfo.c

```diff
@@ -111,9 +112,19 @@ int mm_mallinfo(FAR struct mm_heap_s *heap, FAR struct mallinfo *info)
 
 		mm_takesemaphore(heap);
 
-		for (node = heap->mm_heapstart[region]; node < heap->mm_heapend[region]; node = (struct mm_allocnode_s *)((char *)node + node->size)) {
+		prev = NULL;
+		for (node = heap->mm_heapstart[region]; node < heap->mm_heapend[region]; prev = node, node = (struct mm_allocnode_s *)((char *)node + node->size)) {
 			mvdbg("region=%d node=%p size=%p preceding=%p (%c)\n", region, node, node->size, (node->preceding & ~MM_ALLOC_BIT), (node->preceding & MM_ALLOC_BIT) ? 'A' : 'F');
 
+			if (prev && (prev->size != (node->preceding & ~MM_ALLOC_BIT))) {
+				mdbg("ERROR: Heap node corruption detected\n");
+				mdbg("Corrupted node addr = 0x%08x size = %u\n", prev, prev->size);
```

Comment:
Previous node is corrupting node, current node is corrupted node.
```
Corrupted node addr : node, size : node->size
Corrupting node addr : prev, size : prev->size
```
Am I right?

Reply:
I think corrupting is correct for this case.

Reply:
Node x got corrupted. This resulting in overflow into node x+1. 
So, the preceding value in node x+1 is wrong. Hence, the corruption is at node x.

So, I think @sunghan-chang is right. 
`Corrupting node addr : prev, size : prev->size`

But we are more concerned by who has caused the corruption during the debug stage. So, I think its better to print details of prev node only.

Reply:
```
Node x got corrupted. This resulting in overflow into node x+1.
So, the preceding value in node x+1 is wrong. Hence, the corruption is at node x.
```
Two cases can make above.
1. size of node x is corrupted.
2. preceding of node x + 1 is corrupted.
So, we should show both. (We don't know what is corrupted among them)

Reply:
But, if node x size is corrupted, it is mostly because of overflow from node x-1.
So, it would be caught in the previous for loop iteration.

Reply:
Right. I totally agree what you say. But it's one of the case.
When node x is ok but node (x+1) is corrupted for preceding value, then it will show the same assert.
What I mean is when we meet `size of node (x) != preceding of node (x+1)` assert
1. size of node (x) is broken, all of node (x+1) is ok. => node (x-1) overflow
2. all of node (x) is ok, preceding of node (x+1) is broken. => node (x) overflow
both cases show the same assert. But it is difficult to know what is broken and who overflows.

So, I think we should print 3 nodes information at assert case.

Reply:
In my word, node (x+1) is next node of node (x). node (x-1) is previous node of node (x).

Reply:
Actually, we always try to find overflow by checking value of preceding variable. Not the size variable. 

1. preceding of node x is broken --> our logs show node x-1 information (address, owner ,etc)
2. preceding of node x+1 is broken --> our logs show node x information

In this way, our logs always show the information of the node which caused the overflow.
Also, pls check my changes in mm.h. I have made "preceding" as first member of the node struct. So, whenever there is a overflow from previous node, then preceding of current node will always get corrupted. So, we check 

`if (prev-> size != node->preceding)`
`      print(prev node details)
`

Reply:
I think only `size` could be corrupted. Then, we can get the same assert by `if (prev-> size != node->preceding`.
But this case we need previous of prev.

Reply:
If you think this is enough, then I will merge this immediately

Reply:
I agreed with @sunghan-chang's opinion.
`size` can be corrupted even if preceding is okay.
So we'd better print/check the 3 nodes.

Reply:
@kishore-sn So, could you let me know conclusion?